### PR TITLE
Feature/decoding

### PR DESCRIPTION
**Pull request by simonjwinwood**
Support for decoding x86 binaries via LLVM's tablegen.

### CODE REVIEW

**Comment by simonjwinwood** [comment]
Supersedes pull request #6

File: decodex86/leanpkg.path [hunk 1]
```diff
@@ -0,0 +1,2 @@
+builtin_path
+path ./src
```
Comment (joehendrix) [inline decodex86/leanpkg.path]: leanpkg.path is automatically generated from leanpkg.toml; I don't think we need it in the repo.

File: decodex86/src/decodex86.lean [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  io.fs.write child.stdin input,
+  io.fs.close child.stdin,
+  -- Read input from child stdout
+  buf ← io.fs.read_to_end child.stdout,
+  io.fs.close child.stdout,
+  -- Wait for process to exit
+  exitv ← io.proc.wait child,
+  when (exitv ≠ 0) $ io.fail $ "process exited with status " ++ repr exitv,
+  -- Decode buffer
+  return (decodex86.from_buffer buf)
```
Comment (joehendrix) [inline decodex86/src/decodex86.lean]: Could you add a comment what this should do?

File: lean/elf.lean [hunk 3]
```diff
@@ -318,33 +326,88 @@ instance (c:elf_class) : elf_file_data (word c) :=
...
-(val : uint32)
+inductive phdr_type : Type
+  | PT_NULL    : phdr_type
+  | PT_LOAD    : phdr_type
+  | PT_DYNAMIC : phdr_type
+  | PT_INTERP  : phdr_type
+  | PT_NOTE    : phdr_type 
+  | PT_SHLIB   : phdr_type
+  | PT_PHDR    : phdr_type
+  | PT_PROC    : uint32 -> phdr_type
```
Comment (joehendrix) [inline lean/elf.lean]: I think we should keep the phdr_type as it was.

File: lean/elf.lean [hunk 4]
```diff
@@ -318,33 +326,88 @@ instance (c:elf_class) : elf_file_data (word c) :=
...
-structure phdr_flags :=
-(val : uint32)
+inductive phdr_flag
+  | PF_X : phdr_flag
+  | PF_W : phdr_flag
+  | PF_R : phdr_flag
+/-  | PF_PROC : word c -> phdr_flag -/
+
+@[reducible]
+def phdr_flags := list phdr_flag
```
Comment (joehendrix) [inline lean/elf.lean]: Change back to way it was.

**Pull request merged**